py/gpu_specter/extract/both.py

```diff
@@ -0,0 +1,149 @@
+import numpy as np
+
+try:
+    import cupy as cp
+    cupy_available = True
+except ImportError:
+    cupy_available = False
+
+def xp_deconvolve(pixel_values, pixel_ivar, A):
+    """Calculate the weighted linear least-squares flux solution for an observed trace.
+
+    Args:
+        pixel_values (ny*nx,): 1D array of pixel values
+        pixel_ivar (ny*nx,): 1D array of pixel inverse variances to use for weighting
+        A (ny*nx, nspec*nwave): projection matrix that transforms a 1D spectrum into a 2D image
+
+    Returns:
+        deconvolved (nspec*nwave): the best-fit 1D array of flux values
+        iCov (nspec*nwave, nspec*nwave): the correlated inverse covariance matrix of the deconvolved flux
+
+    """
+    try:
+        xp = cp.get_array_module(A)
+    except NameError:
+        #- If the cupy module is unavailble, default to numpy
+        xp = np
```

Comment:
This is a common pattern used in multiple places.  Consider moving this into `gpu_specter.util.get_array_module`, which would call `cupy.get_array_module` if available, or otherwise return numpy if cupy wasn't imported.

Reply:
I moved this over as you suggested. Does that look okay to you?